nucypher/blockchain/eth/currency.py

```diff
@@ -0,0 +1,197 @@
+
+    def __init__(self, value: Union[int, float, str], denomination: str):
+
+        # Calculate smallest denomination and store it
+        wrapped_denom = self.__denominations[denomination]
+
+        # Validate Early
+        if '.' in str(value):
+
+            _, fraction = str(value).split('.')
+            if len(fraction) > self.__decimals:
+                raise ValueError("Cannot initialize with fractional wei value")
+
+            if wrapped_denom == 'wei':
+                raise ValueError("Cannot initialize with fractional wei value")
+
+        self.__value = currency.to_wei(number=value, unit=wrapped_denom)
+
+    @classmethod
+    def from_nu_wei(cls, value: int):
+        return cls(value, denomination='NUWei')
+
+    @classmethod
+    def from_tokens(cls, value: Union[int, float, str]):
+        return cls(value, denomination='NU')
+
+    def to_tokens(self) -> Decimal:
+        """Returns an decimal value of NU"""
+        return currency.from_wei(self.__value, unit='ether')
+
+    def to_nu_wei(self) -> int:
+        """Returns an int value in NU-Wei"""
+        return int(self.__value)
+
+    def __eq__(self, other) -> bool:
+        return int(self) == int(other)
+
+    def __radd__(self, other) -> int:
+        return int(self) + other
+
+    def __add__(self, other) -> 'NU':
+        return NU(int(self) + int(other), 'NUWei')
+
+    def __sub__(self, other) -> 'NU':
+        return NU(int(self) - int(other), 'NUWei')
+
+    def __rmul__(self, other) -> int:
+        return int(self) * int(other)
+
+    def __mul__(self, other) -> 'NU':
+        return NU(int(self) * int(other), 'NUWei')
+
+    def __floordiv__(self, other) -> 'NU':
+        return NU(int(self) // int(other), 'NUWei')
+
+    def __gt__(self, other) -> bool:
+        return int(self) > int(other)
+
+    def __ge__(self, other) -> bool:
+        return int(self) >= int(other)
+
+    def __lt__(self, other) -> bool:
+        return int(self) < int(other)
+
+    def __le__(self, other) -> bool:
+        return int(self) <= int(other)
+
+    def __int__(self) -> int:
+        """Cast to smallest denomination"""
+        return int(self.to_nu_wei())
+
+    def __repr__(self) -> str:
+        r = f'{self.__symbol}(value={str(self.__value)})'
+        return r
+
+    def __str__(self) -> str:
+        return f'{str(self.to_tokens())} {self.__symbol}'
+
+
+class Stake:
+    """
+    A quantity of tokens, and staking time-frame for one stake for one miner.
+    """
+
+    __ID_LENGTH = 16
+
+    def __init__(self,
+                 owner_address: str,
+                 index: int,
+                 value: NU,
+                 start_period: int,
+                 end_period: int):
+
+        # Stake Info
+        self.owner_address = owner_address
+        self.index = index
+        self.value = value
+        self.start_period = start_period
+        self.end_period = end_period
+        self.duration = self.end_period - self.start_period
```

Comment:
This implies that if I want to stake for a single period, the duration is 0. We may want to either add 1 here or require that `end_period` is the first period after the end of the stake. Alternatively (and I like this option better), is to get rid of `end_period` as a constructor argument, and change it to `duration`.

Reply:
Adding one seems like a simple solution.   I was compelled to accept end_period as a constructor argument, because an on-chain staking record consists of `(start_period, end_period, stake_value)`

Reply:
:heavy_check_mark: